### PR TITLE
Integration tests in TypeScript

### LINKED ISSUE #2267: Use TypeScript for the integration test suite
## Goals

The integration test harness that we recently merged is implemented as vanilla JavaScript. If we wrote this using TypeScript we would get the benefit of testing the types that Realm JS publishes.

## Expected Results

All the tests should be implemented in TypeScript.

## Version of Realm and Tooling

- Realm JS SDK Version: N/A
- Node or React Native: N/A
- Client OS & Version: N/A
- Which debugger for React Native: N/A

### PR DESCRIPTION
**Pull request by kraenhansen**
## What, How & Why?

This closes #2267 by converting all the JavaScript files in the integration test suite to TypeScript.

It also introduces types for two undocumented functions.

## ☑️ ToDos
* [ ] Add tslinting to the tests.
* [ ] Make an attempt at ditching the `node` globals from the test project.
* [x] Go over the README.md files to ensure nothing changed that which was already documented. 
* [x] 📝 Changelog entry
* [x] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

### CODE REVIEW

File: integration-tests/README.md [hunk 1]
```diff
@@ -74,41 +66,32 @@ npm test
...
 
 Tests will have access to the following globals:
 
 - [the Mocha hook globals](https://mochajs.org/#hooks) (define, it, after, before, etc.).
 - `Realm` the Realm constructor.
-- `fs` the lowest common denominator of the [`fs-extra`](https://www.npmjs.com/package/fs-extra) and
-  [`react-native-fs`](https://www.npmjs.com/package/react-native-fs) APIs.
-- `path` the lowest common denominator of the Node.js path interface and a
-  [node-independent implementation of Node's path](https://www.npmjs.com/package/path-browserify) module.
+- An `UndocumentedRealm` which extends the `Realm` constructor with some methods which are not documented or considered a part of the public Realm JS API.
```
Comment (nirinchev) [inline integration-tests/README.md]: While there are undocumented methods/properties, we should aim to have types for all methods, including the private ones.

Comment (kraenhansen) [reply integration-tests/README.md]: I don't disagree - but I don't know if we have consensus on that: Would @kneth or @bmunkholm object against that?

Comment (kneth) [reply integration-tests/README.md]: It's fine with me. You are basically using `UndocumentedRealm` as a box with the label "Nothing important to see".

File: integration-tests/tests/src/dynamic-schema-updates.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
```
Comment (nirinchev) [inline integration-tests/tests/src/dynamic-schema-updates.ts]:
```suggestion
// Copyright 2019 Realm Inc.
```

File: integration-tests/tests/src/index.ts [hunk 3]
```diff
@@ -46,5 +47,9 @@ describe(global.title, () => {
 
 afterEach(() => {
     // Remove all Realm files in the default directory
-    Realm.clearTestState();
+    if ("clearTestState" in Realm) {
+        Realm.clearTestState();
```
Comment (kneth) [inline integration-tests/tests/src/index.ts]: Why don't you have to `(Realm as any).clearTestState()`?

Comment (kraenhansen) [reply integration-tests/tests/src/index.ts]: Because I updated the types of Realm JS to include the private `clearTestState` method :-)

**Pull request merged**